### PR TITLE
Add the possibility to skip dates for recurring events

### LINKED ISSUE #23: Add the possibility to skip dates for recurring events
We're an open source community and started today to evaluate this script. Thank you very much for open sourcing this tool!

Is it already planned to skip dates for recurring events (e.g. if one meeting is canceled due to various reasons)?

I've seen that [yaml2ical](https://opendev.org/opendev/yaml2ical/) already provides this feature, but we'd like to stick to this tool.

Thank you in advance,
Eduard

### PR DESCRIPTION
**Pull request by itrich**
Close #23 .

This PR introduces the sub-key `except_on` to exclude dates from a recurrent event.

As for now, excluding specific times is not yet supported as I struggled with getting timezones right. Maybe someone can give me a hint on how to provide the correct timezone information to the ics file? I'm happy to learn!

Signed-off-by: Eduard Itrich <eduard@itrich.net>

### CODE REVIEW

File: yaml2ics.py [hunk 1]
```diff
@@ -90,6 +90,15 @@ def event_from_yaml(event_yaml: dict, tz: tzinfo = None) -> ics.Event:
             )
         )
 
+        if "except_on" in repeat:
+            for exdate in repeat.get("except_on"):
```
Comment (stefanv) [inline yaml2ics.py:94]:
```suggestion
            for exdate in repeat["except_on"]:
```

File: yaml2ics.py [hunk 2]
```diff
@@ -90,6 +90,15 @@ def event_from_yaml(event_yaml: dict, tz: tzinfo = None) -> ics.Event:
...
+                event.extra.append(
+                    ics.ContentLine(
```
Comment (stefanv) [inline yaml2ics.py:96]: Should this not be combined into one line? The [example from the RFC](https://www.kanzaki.com/docs/ical/exdate.html) is:

```
  EXDATE:19960402T010000Z,19960403T010000Z,19960404T010000Z
```

**Review by stefanv — COMMENTED** [review]
We need some tests, and conformity with the RFC. Otherwise looks good!

Please also test generated calendars with Google Calendar to ensure that everything works as intended.

**Comment by stefanv** [comment]
Thanks for this contribution, @itrich!

**Comment by itrich** [comment]
I'm still struggling with excluding specific times as the timezone needs to be provided explicitily.
 ```
EXDATE:20220630T230000
```
must be
```
EXDATE;TZID="UTC-07:00":20220630T230000
```
but I'm wondering how to achieve this.

Moreover, if only a date is given, the timestamp needs to be completely dropped.

**Comment by itrich** [comment]
From my side, this PR is ready to be reviewed. Exceptions can now be either a `date` (and thus excluding all occurances on this day) or a `datetime` which only excludes this specific occurance. The timezone (if given) is added as param TZID.

Maybe someone can hint me to a better approach to solve this rather long block:
```
            if tz:
                event.extra.append(
                    ics.ContentLine(
                        name="EXDATE",
                        params={"TZID": [str(ics.Timezone.from_tzinfo(tz))]},
                        value=",".join(exdates),
                    )
                )
            else:
                event.extra.append(
                    ics.ContentLine(
                        name="EXDATE",
                        value=",".join(exdates),
                    )
                )
```

**Comment by rkdarst** [comment]
This is great! I will try to have a look at it in the next few days and (unfortunately for me) also test in Outlook calendar in addition to Google calendar. This was a very requested feature and I couldn't figure it out in a reasonable time.

A quick look at the code looks very reasonable to me (not that I know much). The big "if" block looks as good as I know how to make it. One thing though: if `tz` is None, then it converts dtstamp to UTC right below. Does this mess up the conversion? I'm not sure the bust solution: should we require explicit timezones (I guess not?). Should we set tz to UTC by default at the top so it always does uniform conversion? I don't know, will need to investigate more when I have a full computer. But this doesn't need to hold back progress!

**Comment by itrich** [comment]
How shall we proceed with this PR? We need the feature downstream and – if possible – i'd love to contribute this change upstream!

**Review by jarrodmillman — APPROVED** [review]

**Pull request merged**